### ISSUE TITLE
Form MIME Type Validation and guessExtension() of MS Viso (*.vsdx) Files

### ISSUE THREAD

**Issue by gomez77** [symfony/symfony#37547]
**Symfony version(s) affected**: 5.1

**Description**  
I use MIME Type validation for *.pdf, *.doc, *.docx and images successfully. But it seems, that the Microsoft Visio file type format isn't correct recognized.
I get the error code, that the "application/octet-stream" file type isn't allowed.

The same is for the for the guessExtension() function.

It seems this file type isn't recognized fully or correct?

**How to reproduce**  
```php
->add('attachement', FileType::class, [
                'label' => 'measure.frm.labels.attachement',
                'attr' => ['placeholder' => 'measure.frm.placeholder.files', 'accept' => '.pdf,.bmp,.tif,.tiff,.jpeg,.jpg,.png,.gif,.vsdx,.docx,.doc'],
                'mapped' => false,
                'required' => false,
                'multiple' => true,
                'constraints' => [
                    new Count([
                        'max' => 5,
                        'maxMessage' => 'measure.frm.validation.attachement.fileCount'
                        ]),
                    new All([
                        new File([
                            'maxSize' => '5120k',
                            'mimeTypes' => [
                                'application/pdf',
                                'application/x-pdf',
                                'image/*',
                                'application/msword',
                                'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
                                'application/vnd.visio',
                                'application/vnd.ms-visio.drawing.main+xml',
                            ],
                            'maxSizeMessage' => 'measure.frm.validation.attachement.fileMaxSize',
                            'mimeTypesMessage' => 'measure.frm.validation.attachement.fileType',
                    ])
                        ])
                    ],
                'translation_domain' => 'forms',
            ])
```

**Comment by gomez77** [comment]
It isn't working in symfony version 5.1.3 too.

Please can somebody review this issue to verify it.

Thanks

**Comment by xabbuh** [comment]
Can you debug to find out what mime type is returned by `Symfony\Component\HttpFoundation\File\File::getMimeType()` (it's called by the `FileValidator`) and try to find out which mime type guesser instance is returned by `MimeTypeGuesser::getInstance()`?

**Comment by gomez77** [comment]
Hello,

Thanks for your support.

It's both: application/octet-stream

I followed this example: https://symfony.com/doc/current/controller/upload_file.html

With this modification because I can't "filter" *.vsdx files in the FORM

```
$originalFilename = pathinfo($file->getClientOriginalName(), PATHINFO_FILENAME);
$safeFilename = $this->slugger->slug($originalFilename);
/** Check if the MIME Type is known, otherwise use the original file type */
$safeFileType = (is_null($file->getMimeType()) OR $file->getMimeType() === "application/octet-stream") ? $file->getClientOriginalExtension() : $file->guessExtension();
$fileName = $safeFilename.'-'.uniqid().'.'.$safeFileType;
```
For me it's seems, that the "new Visio file format" isn't recognized?

Thanks

**Comment by xabbuh** [comment]
And what kind of `MimeTypeGuesserInterface` instance is used?

**Comment by gomez77** [comment]
The example didn't use a special `MimeTypeGuesserInterface` ?

This is the code

```
<?php

namespace App\Service;

use Symfony\Component\HttpFoundation\File\Exception\FileException;
use Symfony\Component\HttpFoundation\File\UploadedFile;
use Symfony\Component\String\Slugger\SluggerInterface;
use Doctrine\ORM\EntityManagerInterface;
use Doctrine\ORM\ORMException;
use App\Entity\FileTbl;

class FileUploadHelper {
    private $targetDirectory;
    private $slugger;
    private $entityManager;

    public function __construct($targetDirectory, SluggerInterface $slugger, EntityManagerInterface $entityManager) {
        $this->targetDirectory = $targetDirectory;
        $this->slugger = $slugger;
        $this->entityManager = $entityManager;
    }

    public function upload(UploadedFile $file) {

        $fileItem = new FileTbl();

        $originalFilename = pathinfo($file->getClientOriginalName(), PATHINFO_FILENAME);
        $safeFilename = $this->slugger->slug($originalFilename);
        /** Check if the MIME Type is known, otherwise use the original file type */
        $safeFileType = (is_null($file->getMimeType()) OR $file->getMimeType() === "application/octet-stream") ? $file->getClientOriginalExtension() : $file->guessExtension();
        $fileName = $safeFilename.'-'.uniqid().'.'.$safeFileType;

        try {
            $file->move($this->getTargetDirectory(), $fileName);

            $fileItem->setName($file->getClientOriginalName());
            $fileItem->setFileName($fileName);
            $fileItem->setSize($file->getSize());
            $fileItem->setExtension($safeFileType);

            $this->entityManager->persist($fileItem);
            $this->entityManager->flush();

            return $fileItem;

        }
        catch (FileException $e) {
            // ... handle exception if something happens during file upload
        }
        catch (ORMException $exception) {
            // handle doctrine exception
        }

    }

    public function getTargetDirectory() {
        return $this->targetDirectory;
    }
}
```

**Comment by xabbuh** [comment]
You will need to check what instance is returned by `MimeTypeGuesser::getInstance()`. It's called internally but you can also call it in your code.

**Comment by gomez77** [comment]
Sorry I tried to dump it with the following in the code above but without access. I got the error message, that I forgott a `use statement`

Please can you support with the code snipped?

Thanks

```
...
use Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesser;
...
public function upload(UploadedFile $file) {
...
dump(MimeTypeGuesser::getInstance());
...
}
```

**Comment by xabbuh** [comment]
Can you show the exact error that you got?

**Comment by gomez77** [comment]
It isn't a real error. For my point of view it is a miss leading behaviour.

Maybe one step back.

I want to validate/ check the real uploaded file types within a form with the help of "constraints" (see my first post).
For pictures and pdf it is working quite well and I got no errors back from the validation!

In the form validation a also allow Visio files
```
...
 'application/vnd.visio',
 'application/vnd.ms-visio.drawing.main+xml',
....
```

If I try to upload a Visio file I got the "error code" from the form that file type of "application/octet-stream" are not allowed.
So it seems the Visio files are not recognized correct or the MS Visio File format isn't registered?

Here is the error message from the Form (sorry in German)
`Der Dateityp "application/octet-stream" ist nicht zulässig. Es sind nur folgende Dateitypen zulässig: "application/pdf", "application/x-pdf", "image/*", "application/msword", "application/vnd.openxmlformats-officedocument.wordprocessingml.document", "application/vnd.visio", "application/vnd.ms-visio.drawing.main+xml"`

The `guessExtension() ` Function detect the Visio File as *.bin extension too

**Comment by xabbuh** [comment]
I meant the error message of this:

> Sorry I tried to dump it with the following in the code above but without access. I got the error message, that I forgott a use statement

I would like to understand which mime type guesser is used in your case and if there is room for improvement.

**Comment by gomez77** [comment]
Sorry. Here is the error code out from the "dev" enviroment

```
ClassNotFoundError
Attempted to load class "MimeTypeGuesser" from namespace "Symfony\Component\HttpFoundation\File\MimeType".
Did you forget a "use" statement for another namespace?
```

**Comment by xabbuh** [comment]
Ah sorry, the code I was giving you doesn't work on 5.1. Can you please debug inside `Symfony\Component\Mime\MimeTypes::guessMimeType()` which concrete guesser is being used?

**Comment by gomez77** [comment]
with this code out from [https://symfony.com/doc/current/components/mime.html](https://symfony.com/doc/current/components/mime.html)
```
public function upload(UploadedFile $file) {
...
$mimeTypes = new MimeTypes();
 dump($mimeTypes->guessMimeType($file));
...
}
```
I got `"application/octet-stream"` back

**Comment by xabbuh** [comment]
I am sorry, but that does not really help. We will need to find out which concrete guesser is used in your case to decide whether that's something we can improve on the Symfony side or if there's nothing we can do about this.

**Comment by gomez77** [comment]
How can I support? The "only" code I can provide is in the messages above. Especially the first one (the symfony form validation).

I fear symfony is not the root cause. Maybe the "PHP MIME type libraries" should be updated. But which libaries are used? Where can we report this issue?

Thanks

**Comment by xabbuh** [comment]
Inside `Symfony\Component\Mime\MimeTypes:: guessMimeType()` Symfony decides which concrete mime type guesser to use. Please debug a bit more what's happening there (which `MimeTypeGuessInterface` the call is delegated to and what happens inside there).

**Comment by gomez77** [comment]
I think I found the root cause while debugging.
`class FileBinaryMimeTypeGuesser implements MimeTypeGuesserInterface` is using the OS command `public function __construct(string $cmd = 'file -b --mime -- %s 2>/dev/null` to get the file mimetype from the OS.

This is the base of all MimeType functions in symfony. If I run it on the command line, I get also "application/octet-stream" back.
So the "magic libary" used by `file` must be updated. I reported the issue to them too.

**Comment by xabbuh** [comment]
Thank you for investigating. Unfortunately there is nothing we can do about that here. So I am going to close this issue. Thank you for understanding.

**Comment by sebastiansulinski** [comment]
I've just came across a similar problem while validating the uploaded file mime type using `File::getClientMimeType()`.
Problem was only occurring when uploading on Windows via Google Chrome or Edge - uploading via Firefox the mime type was guessed correctly. I was uploading `.doc` document and in both Chrome and Edge I was getting `application/octet-stream` wheres Firefox was giving me `application/vnd.openxmlformats-officedocument.wordprocessingml.document`.

When trying the same on the Mac - Chrome was guessing the mime fine so it is Windows related.
I then went ahead and used `mime_content_type` to get the mime type of the file uploaded via Chrome and Edge on Windows and in both cases I got `application/msword` so evidently there is something not quite right with the mime guesser.
